valhalla/midgard/util.h

```diff
@@ -27,13 +27,21 @@
 namespace valhalla {
 namespace midgard {
 
-// Holds a range plus a default value for that range
+// Holds a named option, its range plus a default value for that range
 template <class T> struct ranged_default_t {
   T min, def, max;
 
+  bool clamped = false;
+  // so we know which costing option this object refers to
+  // important when we clamp the value
+  char* option = "";
+
   // Returns the value snapped to the default if outside of the range
-  T operator()(const T& value) const {
+  T operator()(const T& value, char* option_name = "") const {
     if (value < min || value > max) {
+      clamped = true;
+      // TODO: remove leading "/"
+      option = option_name;
```

Comment:
Turns out warning the user when out-of-range costing options were clamped to defaults is actually quite a big change potentially. #

This is just a suggestion: in the `JSON_PBF_RANGED_DEFAULT` macro we pass the option name and in this struct we flag it as clamped if that's what we're doing. Then later in the processing we could go over all such `ranged_default_t` objects and generate a warning for each that has `clamped = true`. There's LOTS of these range objects, so it'll be a bit ugly, unless we collect them even more in smth like `BaseCostingOptionsConfig`.